constraints_lang.md

```diff
@@ -0,0 +1,311 @@
+
+```
+travel_length = 10
+tolerance = 0.03
+my_radious > 0
+2 * pi * my_radius < travel_length
+maximize(my_radius)
+```
+
+By adding `maximize(my_radius)`, a specific value can be computed for `my_radius` even though it's not fully constrained. This is useful for displaying geometry that would only be partially constrained otherwise.
+
+Resolution strategies are only used if values are not fully constrained otherwise. You can think of them as optional constraints. This is useful for library authors while writing constraint functions. Depending on what a constraint function is applied to, the result may or may not be fully constrained. Resolution strategies allow for a way to give hints to the runtime system that aren't strict requirements. I.e. resolution strategies will never cause geometry to be over-constrained. But they can help allow unconstrained or partially constrained geometry to become fully constrained.
+
+> Half-baked: Initially, the only resolution strategies will be `minimize` and `maximize` to deal with inequalities. In future versions, we would allow users to define their own resolution strategies. The gist is that they need to define an implementation for a trait like this.
+
+```
+trait Resolve<T> {
+  fn resolve(cs: Set<Constraint>) -> T
+}
+```
+
+> It's parameterized with a type since a resolution strategy may only apply to numbers, for example, not vectors or lines. This is a very under-baked idea because if it's possible to define resolution strategies for composite types, multiple strategies could apply. How would we determine which ones to apply?
+### Applying pure functions and constraints
+
+At the top level of code, there's no use in applying pure functions.
+
+```
+dot(v1, v2)
+```
+
+This has no effect because `dot()` is a pure function. It computes the dot product of two vectors and returns it. The return value is ignored.
+
+On the other hand, constraints are useful at the top level.
+
+> Half-baked: Pure functions can use `constraint`s internally. However, any constraints built up do not escape the scope of the pure function. For this reason, a rule of thumb may be to only apply constraints to local variables, not parameters. But this is not a strict requirement. Any constraints applied to parameters are simply discarded before returning to the caller. Because it's defined in a pure function with `fn`, they follow the same lexical scoping rules. This is an under-baked idea because although this all sounds nice -- we don't want to create a split in the language where you can't use constraint code when writing a pure function -- it's unclear how this would be implemented. In theory, it could be as simple as cloning any input arguments, but I'm not sure how that works with the implementation discussed later.
+
+### Displaying Geometry
+
+Any geometry that is fully constrained will be shown by default. A user can opt-out of this using:
```

Comment:
I could use a little more explanation around what it means to be *fully constrained* in this paradigm's sense. I think it means in the mathematical sense that the equation we're we're solving for has more than one solution, but that seems distinct from the CAD meaning of the term, in which one or more positions are "movable". I think they are the same underlying idea in both senses, but in CAD you typically display "not fully constrained" sketch geometry (traditionally black geometry is constrained and blue geometry could be moved around).

Reply:
I don't know what you mean. Maybe I need to read up on how CAD software uses the term. I intended "fully constrained" to mean that a variable, like the X-coordinate of a point, has a unique and computable value. If this isn't the case, we cannot know where to display it on the screen.

Reply:
@jtran Under-constrained sketches are always displayed in CAD, with the program leveraging construction values as guesses (rendered in blue not black as @franknoirot said) or making sure the points don't fly away when a constraint is applied (likely doing some sort of distance minimization while meeting the constraints)

Reply:
Here's an example of a sketch that has two segments that are constrained (because I made the lower-right point coincident with the origin) and two that are not.

![Screenshot 2024-10-22 at 8 52 06 PM](https://github.com/user-attachments/assets/62cbeda7-0fca-4c6a-a0be-621620e3d813)